url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/io/image/ImagePersistence.java

```diff
@@ -0,0 +1,693 @@
+                    bimg.setRGB(x, y, 
+                      (VSDK.signedByte2unsignedInteger(p.r)) * 256 * 256 +
+                      (VSDK.signedByte2unsignedInteger(p.g)) * 256 +
+                      (VSDK.signedByte2unsignedInteger(p.b))
+                    );
+                }
+            }
+
+            FileOutputStream fos = new FileOutputStream(fd);
+
+            ImageIO.write(bimg, "gif", fos);
+
+            // OLD DEPRECATED API, DO NOT USE!
+            //com.sun.image.codec.jpeg.JPEGImageEncoder jpeg;
+            //jpeg = com.sun.image.codec.jpeg.JPEGCodec.createJPEGEncoder(fos);
+            //jpeg.encode(bimg);
+
+            fos.close();
+        }
+        catch ( Exception e ) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+    This method writes the contents of the specified image to a file in 
+    binary RGB PPM format (i.e. P6 PPM sub-format). Returns true if everything
+    works fine, false if something fails, like a permission access denied
+    or if storage device runs out of space.
+    */
+    public static boolean exportPPM(File fd, Image img)
+    {
+        try {
+            BufferedOutputStream writer;
+            FileOutputStream fos = new FileOutputStream(fd);
+            writer = new BufferedOutputStream(fos);
+
+            String line1 = "P6\n";
+            String line2 = "# Image generated by VitralSDK (http://vitral.sf.net)\n";
+            String line3 = img.getXSize() + " " + img.getYSize() + "\n";
+            String line4 = "255\n";
+            byte arr[];
+
+            arr = line1.getBytes();
+            writer.write(arr, 0, arr.length);
+            arr = line2.getBytes();
+            writer.write(arr, 0, arr.length);
+            arr = line3.getBytes();
+            writer.write(arr, 0, arr.length);
+            arr = line4.getBytes();
+            writer.write(arr, 0, arr.length);
+
+            RGBPixel p;
+            int x = 0, y = 0;
+            for ( y = 0; y < img.getYSize(); y++ ) {
+                for ( x = 0; x < img.getXSize(); x++ ) {
+                    p = img.getPixelRgb(x, y);
+                    writer.write(p.r);
+                    writer.write(p.g);
+                    writer.write(p.b);
+                }
+            }
+
+            writer.flush();
+            writer.close();
+            fos.close();
+        }
+        catch (Exception e) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+    This method writes the contents of the specified image to a file in 
+    binary RGB uncompressed BMP format. Returns true if everything
+    works fine, false if something fails, like a permission access denied
+    or if storage device runs out of space.
+    */
+    public static boolean exportBMP(File fd, Image img)
+    {
+        try {
+            BufferedOutputStream writer;
+            FileOutputStream fos = new FileOutputStream(fd);
+            writer = new BufferedOutputStream(fos);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
Analysis of this package determined that this line of code contains a resource that might not have closed properly. A resource leak can slow down or crash your system.

**Fix**
Consider closing the following resource in a try-finally block: *writer*. The resource is referenced at lines 512, 521, 522, 523, 524, 527, 528, 529, 530, 531, 532, 533, 534, 535, 536, 537, 545, 546, 547, 551. The resource is closed at lines 552, 553. There are other execution paths that do not contain closure statements, for example, when BufferedOutputStream.write throws an exception.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).